lib/jerboa/client/worker.ex

```diff
@@ -96,161 +103,220 @@ defmodule Jerboa.Client.Worker do
     end
   end
 
-  @spec request_permission(state, peers :: [Client.ip, ...],
-    retries_left :: pos_integer) :: {result :: term, state}
-  def request_permission(state, peer_addrs, retries_left) do
-    {result, new_state} =
-      state
-      |> Protocol.create_perm_req(peer_addrs)
-      |> send_req()
-      |> Protocol.eval_create_perm_resp()
-    with :retry <- result,
-         n when n > 0 <- retries_left do
-      Logger.debug fn ->
-        "Received error create permission response, retrying.."
+  @spec allocate_response_handler :: Transaction.handler
+  defp allocate_response_handler do
+    fn params, creds, relay ->
+      case Allocate.eval_response(params, creds) do
+        {:ok, relayed_address, lifetime} ->
+          Logger.debug fn ->
+            "Received success allocate reponse, relayed address: " <>
+              Client.format_address(relayed_address)
+          end
+          new_relay =
+            relay
+            |> Map.put(:address, relayed_address)
+            |> Map.put(:lifetime, lifetime)
+            |> update_allocation_timer()
+          reply = {:ok, relayed_address}
+          {reply, creds, new_relay}
+        {:error, reason, new_creds} ->
+          Logger.debug fn ->
+            "Error when receiving allocate response, reason: #{inspect reason}"
+          end
+          reply = {:error, reason}
+          {reply, new_creds, relay}
       end
-      request_permission(new_state, peer_addrs, n - 1)
-    else
-      :ok ->
-        {result, update_permissions(new_state, peer_addrs)}
-      _ ->
-        {result, new_state}
     end
   end
 
-  @spec setup_logger_metadata(state) :: any
-  defp setup_logger_metadata(%{socket: socket, server: server}) do
-    {:ok, port} = :inet.port(socket)
-    metadata = [jerboa_client: "#{inspect self()}:#{port}",
-                jerboa_server: Client.format_address(server)]
-    Logger.metadata(metadata)
+  @spec refresh_response_handler :: Transaction.handler
+  defp refresh_response_handler do
+    fn params, creds, relay ->
+      case Refresh.eval_response(params, creds) do
+        {:ok, lifetime} ->
+          Logger.debug fn ->
+            "Received success refresh reponse, new lifetime: " <>
+              "#{lifetime}"
+          end
+          new_relay =
+            relay
+            |> Map.put(:lifetime, lifetime)
+            |> update_allocation_timer()
+          {:ok, creds, new_relay}
+        {:error, reason, new_creds} ->
+          Logger.debug fn ->
+            "Error when receiving refresh response, reason: #{inspect reason}"
+          end
+          reply = {:error, reason}
+          {reply, new_creds, relay}
+      end
+    end
   end
 
-  @spec update_permissions(state, [Client.ip, ...]) :: state
-  defp update_permissions(state, peer_addrs) do
-    state
-    |> remove_permissions(peer_addrs)
-    |> add_permissions(peer_addrs)
+  @spec create_perm_response_handler :: Transaction.handler
+  def create_perm_response_handler do
```

Comment:
Do we really need to return an unnamed function? Why can't we just use `&create_perm_response_handler/3`? This comment applies to all those handlers. If we would skip this `fn` wrapper, we get less nested function.